inst/include/dust/random/gamma.hpp

```diff
@@ -0,0 +1,225 @@
+#ifndef DUST_RANDOM_GAMMA_HPP
+#define DUST_RANDOM_GAMMA_HPP
+
+#include <algorithm>
+#include <cmath>
+#include <stdexcept>
+
+#include "dust/random/generator.hpp"
+#include "dust/random/numeric.hpp"
+#include "dust/random/exponential.hpp"
+#include "dust/random/uniform.hpp"
+#include "dust/random/normal.hpp"
+
+// For 'shape >= 1' follows Ahrens, J. H. and Dieter, U. (1982). Generating gamma variates by
+// a modified rejection technique. For '0 < shape < 1' uses Ahrens, J. H. and Dieter,
+// U. (1974). Computer methods for sampling from gamma, beta, Poisson and binomial distributions.
+// and follows the R implementation: https://github.com/wch/r-source/blob/trunk/src/nmath/rgamma.c
+namespace dust {
+namespace random {
+namespace {
+
+template <typename real_type>
+void gamma_validate(real_type a, real_type b) {
+  if (a < 0.0 || b < 0.0) {
+    char buffer[256];
+    snprintf(buffer, 256,
+             "Invalid call to gamma with a = %g, b = %g",
+             a, b);
+    dust::utils::fatal_error(buffer);
+  }
+}
+
+// Generate gamma random number via GS, p 228 of reference.
+template <typename real_type, typename rng_state_type>
+real_type gamma_gs(rng_state_type& rng_state, real_type a) {
+  real_type b = 1.0 + std::exp(-1) * a;
+  while (TRUE) {
+    real_type p = b * uniform<real_type>(rng_state, 0, 1);
+    if (p >= 1.0) {
+      real_type x = -log((b - p) / a);
+      if (exponential_rand<real_type>(rng_state) >= (1.0 - a) * log(x))
+        return x;
+    } else {
+      real_type x = exp(log(p) / a);
+      if (exponential_rand<real_type>(rng_state) >= x)
+        return x;
+    }
+  }
+}
+
+// Generate gamma random number via GD, p 53 of reference.
+template <typename real_type, typename rng_state_type>
+real_type gamma_gd(rng_state_type& rng_state, real_type a) {
+  const real_type q1 = 0.04166669;
```

Comment:
Codecov is complaining about the method complexity here. It could certainly be refactored but at the moment is very close to the R source implementation (minus a couple of superfluous lines) which is perhaps an advantage in other ways (can easily be verified as equivalent to that implementation).